iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp

```diff
@@ -297,23 +299,39 @@ struct ConvertHLOToLinalgOnTensorsPass
     MLIRContext *context = &getContext();
 
     auto typeConverter = mhlo::createHloToLinalgSignedIntegerConverter();
-    populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
-                                                   patterns);
+    if (directHloClientLowering) {
+      // NOTE: not using corresponding setupHLOToFlowPatterns because the entire
+      // HLO dialects are marked illegal by this pass.
+      // TODO: Wonky.
+      populateHLOToFlowPatterns(context, patterns);
+      chlo::PopulateDecomposeChloPatterns(context, &patterns);
+      populateHLOBroadcastingToLinalgPatterns(context, *typeConverter,
+                                              patterns);
+      populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
+                                                     patterns);
+    } else {
+      // Legacy: assumes that HLO client -> HLO has been run previously.
+      populateHLOToLinalgOnTensorsConversionPatterns(context, *typeConverter,
+                                                     patterns);
+    }
+
     patterns.insert<PadTensorOpConversion>(context);
 
     ConversionTarget target(getContext());
     target.addIllegalDialect<mhlo::MhloDialect>();
 
+    if (directHloClientLowering) {
+      // Conversions to also legalize the HLO client dialect.
+      target.addIllegalDialect<chlo::HloClientDialect>();
```

Comment:
Would it make sense to always include this? I would imagine we are hosed if HloClient operations exist so it may be worth bailing out, even if using the legacy path.

Reply:
For this and the comment above, I was trying to keep this commit NFC and easy to strip out the old without thinking in ~a week. You would be shocked what we currently allow through the front-door so being conservative. One of the next commits will install a better front-door guard.

I may need to do some additional special casing before this lands and wanted to avoid intertwined things to ease my next steps.